### ISSUE TITLE
ProxyHeaderParser not fully supporting multiple reverse proxies

### ISSUE THREAD

**Issue by diamondq** [micronaut-projects/micronaut-core#4571]
TL;DR: Incorrect information is returned when parsing `Forwarded` headers placed by multiple reverse proxies. This causes problems generating the public URL for a Micronaut application (and really breaks things like OAuth security redirects).

The [RFC 7239 Forwarded HTTP Extension](https://tools.ietf.org/html/rfc7239#section-4) isn't super clear with regards to having multiple Forwarded field-values when multiple proxies are involved. The spec indicates in Section 4:

`A proxy server that wants to add a new "Forwarded" header field value can either append it to the last existing "Forwarded" header field after a comma separator.`

`Each field-value is a semicolon-separated list; this sublist consists of parameter-identifier pairs.`

Unfortunately, all their examples for multiple field-values don't have multiple parameter-identifier pairs. The Micronaut code in ProxyHeaderParser interpreted it to mean that the primary separator is the semi-colon, and then within the `for` parameter, there could be a multiple values separated by comma. However, it's actually the other way around.

### Steps to Reproduce

1. If a HttpRequest with a header of `Forwarded: for=192.168.0.55;proto=https;host=example.com,for=172.19.0.1;proto=http` is passed to ProxyHeaderParser

### Expected Behaviour

Based on the spec, it should break it apart by comma first, into two parts, one for each proxy (`for=192.168.0.55;proto=https;host=example.com` and `for=172.19.0.1;proto=http`), and then for each part, break it apart by a semi-colon. Unfortunately, the API for ProxyHeaderParser doesn't provide a way to get the parameters for each component. But if all the fields were arrays, then we'd have something like

```
forwardedFor=[192.168.0.55,172.19.0.1]
forwardedProto=[https,http]
forwardedHost=[example.com,null]
forwardedPort=[null,null]
```

If we don't want to change the exposed API, then we should only parse the first part/proxy, since that should be the outer proxy, which would contain the information needed to regenerate the public URL.

### Actual Behaviour

It breaks it up by semi-colon, and then only by comma for the `for` element. However, since the `for` element in this example is NOT next to each other, you end up with

```
forwardedFor=[192.168.0.55]
forwardedProto=http
forwardedHost=null
forwardedPort=null
```

since it would try to parse "host=example.com,for=172.19.0.1" and that's not a valid simple parameter.

The actual result is wrong in three ways. First, it's missing the `host`, which is used to generate public URLs. And two, the `proto` is inconsistent, as it's now http, but the first proxy was https. Lastly, but less critical, it's also missing the final `for` value.

### Environment Information

- **Operating System**: Windows (but shouldn't matter)
- **Micronaut Version:** 2.1.3
- **JDK Version:** 11 (but shouldn't matter)

**Comment by graemerocher** [comment]
Seems like you have provided a good analysis of the problem. Thanks! Would you be interested in submitting a PR?
